telethon/helpers.py

```diff
@@ -95,17 +95,24 @@ def strip_text(text, entities):
     return text
 
 
-def retry_range(retries):
+def retry_range(retries, force_retry=True):
     """
     Generates an integer sequence starting from 1. If `retries` is
     not a zero or a positive integer value, the sequence will be
     infinite, otherwise it will end at `retries + 1`.
     """
-    yield 1
+
+    # We need atleast one iteration even if the retries are 0
+    # when force_retry is True. 
+    if retries == 0 and force_retry:
+        yield 1
+
+    # If retries are non 0 then iterate
     attempt = 0
     while attempt != retries:
         attempt += 1
-        yield 1 + attempt
+        yield attempt
```

Comment:
This will yield 1 twice. Perhaps moving `attempt` a bit earlier would make this cleaner:

```python
attempt = 0
if retries == 0 and force_retry:
    attempt += 1
    yield attempt

# If retries are non 0 then iterate
while ...:
```

Reply:
it will not yield 1, 2 times
as `attempt != retries` will be false at the first stage and thus never yield an extra 1.

Reply:
For this snippet:

```python
print('0 retries, force', list(retry_range(0)))
print('1 retries, force', list(retry_range(1)))
print('2 retries, force', list(retry_range(2)))
```

The old code prints:

```
print('0 retries, force', list(retry_range(0)))
print('1 retries, force', list(retry_range(1)))
print('2 retries, force', list(retry_range(2)))
```

However, with the current changes, the code now prints (including `force=False`):

```
0 retries, force [1]
1 retries, force [1]
2 retries, force [1, 2]
0 retries, no force []
1 retries, no force [1]
2 retries, no force [1, 2]
```

Note that this behavior is not correct when forcing to retry, since it yields 1 twice.

Reply:
Ha, I was right! %-)